### PR TITLE
feat(discriminator): Add Discriminator support for Interfaces

### PR DESCRIPTION
**Pull request by karatakis**
**Summary:**  
Enable Discriminator for Interface types

**Issue Reference(s):**  
Fixes https://github.com/tailcallhq/tailcall/issues/2961

**Build & Testing:**

- [X] I ran `cargo test` successfully.
- [X] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [X] I have added relevant unit & integration tests.
- [X] I have updated the [documentation] accordingly.
- [X] I have performed a self-review of my code.
- [X] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

**Comment by karatakis** [comment]
depends on https://github.com/tailcallhq/tailcall/pull/2963

File: src/core/config/config.rs [hunk 1]
```diff
@@ -68,6 +68,11 @@ pub struct Config {
     #[serde(default, skip_serializing_if = "is_default")]
     pub unions: BTreeMap<String, Union>,
 
+    ///
+    /// A map of all the interface types in the schema.
+    #[serde(default, skip_serializing_if = "is_default")]
+    pub interfaces: BTreeMap<String, Interface>,
```
Comment (meskill) [inline src/core/config/config.rs:74]: from Config perspective any type that is used as base for other type is interface. Please check https://github.com/tailcallhq/tailcall/blob/d68f44cc9477db78fc3cec4e373fe5b2f3afb5d3/src/core/config/config_module.rs#L82

I'm not sure we need this change to the config - interface types could be checked from the config_module

File: src/core/ir/eval.rs [hunk 2]
```diff
@@ -68,20 +68,42 @@ impl IR {
...
+                                    } else {
+                                        Err(Error::ExprEval(format!(
+                                            "Can't find mapped key: {}.",
+                                            key
+                                        )))
+                                    }
+                                } else {
+                                    Err(Error::ExprEval(
+                                        "Mapped key must be either string or array value."
+                                            .to_owned(),
```
Comment (meskill) [inline src/core/ir/eval.rs:96]: better move this to separate function that can be called recursively. It'll lower code duplication and will be able to handle lists inside lists

**Review by meskill — CHANGES_REQUESTED** [review]

File: src/core/config/config_module.rs [hunk 3]
```diff
@@ -31,19 +31,22 @@ struct Cache {
     input_types: HashSet<String>,
     output_types: HashSet<String>,
     interface_types: HashSet<String>,
+    interfaces_types_map: HashMap<String, HashSet<String>>,
```
[the remark this answers is not in the document]
Comment (karatakis) [reply src/core/config/config_module.rs:34]: its completely different each one

let's say you have 
```gql

type A implements T1 { }
type B implements T2 & T3 { }
type C implements T4 { }

interface T1 { }
interface T2 { }
interface T3 { }
interface T4 implements T2 { }
```

**interface_types** is going to be `[T1, T2, T3, T4]`

**interfaces_types_map** is going to be
```json
{
  "T1": ["A"],
  "T2": ["B", "C"],
  "T3": ["B"],
  "T4": ["C"],
}
```


I can use map => keys to replace **interface_types**

File: src/core/config/config.rs [hunk 4]
```diff
@@ -523,6 +523,55 @@ impl Config {
...
+                        types_set_local = types_set_local.merge_right(types_set_inner);
+                    }
+                    None => {
+                        types_set_local.insert(type_name.to_string());
+                    }
+                }
+            }
+
+            types_set_local
+        }
```
[the remark this answers is not in the document]
Comment (karatakis) [reply src/core/config/config.rs:555]: Good catch, infinite recursion have not been tested

[the remark this answers is not in the document]
Comment (karatakis) [reply src/core/config/config.rs:555]:
```gql
type Query {
  data: A
}

type A implements T1 & T3 & T4 {
  id: ID!
  foo: String
}

type B implements T2 {
  id: ID!
  bar: String
}

interface T1 implements T4 & T3 {
  foo: String
}

interface T2 {
  bar: String
}

interface T3 implements T1 & T4 {
  foo: String
}

interface T4 implements T3 & T1 {
  foo: String
}
```

If you use this schema you will get an error indicating that Interfaces cannot reference each other, so there is no fear of infinity toops

[the remark this answers is not in the document]
Comment (karatakis) [reply src/core/config/config.rs]: @meskill its hard to follow the same approach as `src/core/config/transformer/nested_unions.rs` because I do not store something in the `Config` object to implement a transformer on `Config`. But I can place it in a separate file. Will this work?

[the remark this answers is not in the document]
Comment (meskill) [reply src/core/config/config.rs]: could you please check that we have a test for this? And please verify the recursion case against json/yaml config also, similar to `tests/execution/test-interface-from-json.md`

about the code itself - I see the issue with the transformer, let's stick with your current approach, it's fine

File: src/core/config/config.rs [hunk 5]
```diff
@@ -510,17 +510,47 @@ impl Config {
...
+                    Some(types_set_inner) => {
+                        let types_set_inner =
+                            recursive_interface_type_merging(types_set_inner, interfaces_types);
+                        types_set_local = types_set_local.merge_right(types_set_inner);
+                    }
+                    None => {
+                        types_set_local.insert(type_name.to_string());
+                    }
+                }
             }
```
Comment (meskill) [inline src/core/config/config.rs:542]: small note about possible optimization here:
we recalculate the types every time even if we have already calculated some subtrees

for this example:
```graphql
interface T1 {}
interface T2 implements T2 {}
interface T3 implements T1 & T2 {}
type A implements T1 & T2 & T3 {}
```

when executing T1 we will calculate that A belongs to T2 through T3, but when we start processing T2 we will rerun all the checks through T3 again.
Hope it's not a big deal =)

**Review by meskill — APPROVED** [review]
